junit-platform-launcher/src/main/java9/org/junit/platform/launcher/listeners/FlightRecordingListener.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2015-2020 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.platform.launcher.listeners;
+
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+import jdk.jfr.Category;
+import jdk.jfr.Event;
+import jdk.jfr.Label;
+import jdk.jfr.Name;
+
+import org.apiguardian.api.API;
+import org.junit.platform.engine.TestExecutionResult;
+import org.junit.platform.launcher.TestExecutionListener;
+import org.junit.platform.launcher.TestIdentifier;
+import org.junit.platform.launcher.TestPlan;
+
+/**
+ * A {@link TestExecutionListener} that generates Java Flight Recorder
+ * events.
+ *
+ * @see <a href="https://openjdk.java.net/jeps/328">JEP 328: Flight Recorder</a>
+ * @since 1.7
+ */
+@API(status = EXPERIMENTAL, since = "1.7")
+public class FlightRecordingListener implements TestExecutionListener {
+
+	private final AtomicReference<TestPlanExecutionEvent> testPlanExecutionEvent;
+	private final Map<String, TestExecutionEvent> testExecutionEventMap;
+
+	public FlightRecordingListener() {
+		this.testPlanExecutionEvent = new AtomicReference<>();
+		this.testExecutionEventMap = new ConcurrentHashMap<>();
+	}
+
+	@Override
+	public void testPlanExecutionStarted(TestPlan plan) {
+		TestPlanExecutionEvent event = new TestPlanExecutionEvent();
+		event.containsTests = plan.containsTests();
+		event.rootContainerIds = plan.getRoots().stream().map(TestIdentifier::getUniqueId).collect(
+			Collectors.joining("#"));
+		testPlanExecutionEvent.set(event);
+		event.begin();
+	}
+
+	@Override
+	public void testPlanExecutionFinished(TestPlan plan) {
+		TestPlanExecutionEvent event = testPlanExecutionEvent.get();
+		event.commit();
+	}
+
+	@Override
+	public void dynamicTestRegistered(TestIdentifier test) {
+		DynamicTestEvent event = new DynamicTestEvent();
+		event.uniqueId = test.getUniqueId();
+		event.displayName = test.getDisplayName();
+		event.commit();
+	}
+
+	@Override
+	public void executionSkipped(TestIdentifier test, String reason) {
+		SkippedTestEvent event = new SkippedTestEvent();
+		event.uniqueId = test.getUniqueId();
+		event.displayName = test.getDisplayName();
+		event.reason = reason;
+		event.commit();
+	}
+
+	@Override
+	public void executionStarted(TestIdentifier test) {
+		TestExecutionEvent event = new TestExecutionEvent();
```

Comment:
❓ Does JFR support explicit nesting of events?

Reply:
Yes, but it is usually good to avoid from a visualization point of view.

This image from JMC shows what happens in the JVM (thread sleep, lock contention etc) using events from JFR. A JUnit test event would be visualized on separate lane next to the JVM event so you can see for what made a test slow, i.e a call to Thread.sleep(). 

Nested events are painted on top of each other, so it can be hard to see. Preferably only the actual test (the leaf nodes) should be represented by the Test event. If you want to have information about test containers/groups, it would be better to separate them into their own events, or perhaps leave them out completely.

https://docs.oracle.com/javase/10/troubleshoot/img/jfr-eventgraph.png

Reply:
> [...] If you want to have information about test containers/groups, it would be better to separate them into their own events, or perhaps leave them out completely.

Good idea.

Candidates to be left out:
- Test Plan
- Dynamic Test Registration
- Test Container

They don't contribute much to a better understanding what happened in a test run. Also, the recorded "Unique Id" field already contains the hierarchical information of a test and its parents.

Reply:
I would leave them out for now. It will reduce the noise. If there is a need later on, they can be added.

Reply:
Reduced the scope to "Test Plan" and "Test" (w/o containers) events for the time being.